File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    tuple<bool, int, int> solve(TreeNode* root)
    {
        if(root == nullptr) return {true, INT_MIN, INT_MAX};
        auto [lst, lmax, lmin] = solve(root->left);
        auto [rst, rmax, rmin] = solve(root->right);
        if(!lst || !rst) return {false, INT_MIN, INT_MAX};
        if(lmax >= root->val) return {false, INT_MIN, INT_MAX};
        if(rmin <= root->val) return {false, INT_MIN, INT_MAX};
        return {true, max(rmax, root->val), min(lmin, root->val)};
    }

    bool isValidBST(TreeNode* root) {
        // find max node of left tree
        // find min node of right tree
        // recurisvely
        auto [res, rmin, rmax] = solve(root);
        return res;
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def cns(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None and subRoot is None:
            return True
        if root is None or subRoot is None:
            return False
        if root.val != subRoot.val:
            return False
        if self.cns(root.left, subRoot.left) and self.cns(root.right, subRoot.right):
            return True
        return False
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None:
            return False
        if self.cns(root, subRoot):
            return True
        if self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot):
            return True
        return False
        